### PR TITLE
feat: add optional submodule initialization for worktrees

### LINKED ISSUE #1187: Worktrees created from monorepos do not initialize git submodules
## Summary

When `git worktree add` creates a new worktree from a monorepo that uses git submodules, the submodule directories in the new worktree are empty. Archon does not run `git submodule update --init` after worktree creation, so any workflow that needs to operate on submodule content will find empty directories.

## Root Cause

`WorktreeProvider.createWorktree()` (`packages/isolation/src/providers/worktree.ts`) calls `git worktree add` without `--recurse-submodules`, and no `git submodule update --init` is run afterward. Git's default `worktree add` behavior creates the `.gitmodules` file but leaves submodule directories empty.

## Expected Behavior

After worktree creation, submodules should be initialized and checked out so that workflows can operate on the full monorepo content.

## Suggested Fix

After `git worktree add` succeeds in `createNewBranch()`, run:
```
git -C <worktreePath> submodule update --init --recursive
```

This should be conditional — only when the repo has a `.gitmodules` file — to avoid unnecessary overhead for repos without submodules. Consider making it configurable via `.archon/config.yaml` (`worktree.initSubmodules: true/false`).

## Related

- #1183 — worktree reuse across different local clones (fixed in #1186)
- Parallel workflow collision in shared worktree paths (see below)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by halindrome**
## Summary

Fixes #1187

Adds a `worktree.initSubmodules` config option that runs `git submodule update --init --recursive` after worktree creation. This ensures workflows operating on monorepos with submodules have access to submodule content in their isolated worktrees.

## Changes

- **`packages/isolation/src/types.ts`**: Add `initSubmodules?: boolean` to `WorktreeCreateConfig`
- **`packages/core/src/config/config-types.ts`**: Add `initSubmodules?: boolean` to `RepoConfig.worktree`
- **`packages/isolation/src/providers/worktree.ts`**: Add `initSubmodules()` method called after `git worktree add` when configured. Checks for `.gitmodules` before running (no overhead for repos without submodules). Non-fatal — logs a warning if submodule init fails.
- **`packages/isolation/src/providers/worktree.test.ts`**: 5 new tests covering: init when enabled, skip when disabled, skip when not configured, skip when no `.gitmodules`, and graceful failure.

## Configuration

```yaml
# .archon/config.yaml
worktree:
  initSubmodules: true  # default: false
```

## Design decisions

- **Opt-in**: Submodule init adds network I/O and time. Repos without submodules pay zero cost.
- **Non-fatal**: Submodule init failure (network issues, auth problems) logs a warning but doesn't block worktree creation. The worktree is still usable for work that doesn't touch submodules.
- **120s timeout**: Submodule init can be slow for large repos with many submodules. The 120s timeout is 4x the standard git operation timeout.
- **Recursive**: Uses `--recursive` to handle nested submodules.

## Test plan

- [x] `bun run type-check` — all packages pass
- [x] `bun run lint` — zero warnings
- [x] `bun run test` — all tests pass (5 new tests for submodule init)
- [ ] Manual: create a repo with submodules, set `worktree.initSubmodules: true`, run a workflow, verify submodule dirs are populated

## Related

- #1183 — worktree reuse across different local clones
- #1188 — git-level worktree adoption bypass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * New worktree.initSubmodules option (defaults to true); worktree creation initializes git submodules when present unless explicitly disabled.

* **Bug Fixes / Behavior**
  * Submodule init failures now surface and can block worktree creation. Absence of .gitmodules skips init; permission errors during .gitmodules access fail fast.

* **Tests**
  * Added coverage for init on/off, missing vs permission access, failure cases, and invocation counting.

* **Documentation / Errors**
  * Updated docs and user-facing error guidance for submodule initialization.

### CODE REVIEW

**Comment by halindrome** [comment]
## QA Round 1 Report

### Critical Issues

| # | Severity | Confidence | Finding | Status |
|---|----------|------------|---------|--------|
| 1 | Critical | 93 | `directoryExists` call outside `try/catch` — non-ENOENT errors (e.g., EACCES) crash worktree creation | **Fixed** in 85ce222 |

The `.gitmodules` access check was outside the non-fatal `try/catch`, meaning a permission error would propagate and abort worktree creation. Fixed by using `fs.access` directly inside its own `try/catch` — ENOENT means no submodules (skip silently), other errors log a warning and skip.

### Important Issues

| # | Severity | Confidence | Finding | Status |
|---|----------|------------|---------|--------|
| 4 | Important | 82 | `copyConfiguredFiles` has stale inline type (missing `initSubmodules`) | Advisory — no functional impact |
| 5 | Important | 85 | `directoryExists` name misleading when used on a file | Resolved by not using it — `initSubmodules` now calls `access()` directly |
| 6 | Important | 80 | 120s timeout is unconfigurable and silent | Advisory — documented in PR description |
| 7 | Important | 85 | Missing test for non-ENOENT error from `.gitmodules` check | **Fixed** — added EACCES test in 85ce222 |

### Advisory Notes

- `copyConfiguredFiles` uses an inline type `{ baseBranch?: string; copyFiles?: string[] }` that doesn't include `initSubmodules`. This is structurally compatible (TypeScript won't error) and `copyConfiguredFiles` doesn't need the field, but it's maintenance debt. Not blocking.
- 120s timeout is 4x the standard 30s used elsewhere. Justified for large submodule trees but not configurable. Could add `worktree.submoduleInitTimeout` in a follow-up.

### Verdict

Critical issue fixed. QA round 1 complete.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by halindrome** [comment]
## QA Round 2 Report

### Findings

| # | Severity | Confidence | Finding | Status |
|---|----------|------------|---------|--------|
| 1 | Important | 82 | Log event name mismatch: `worktree.submodules_initialized` doesn't pair with `worktree.submodule_init_failed` per CLAUDE.md convention | **Fixed** in 9ebbc8f — renamed to `worktree.submodule_init_completed` |

### Verified Clean

- **Non-fatal contract**: Both `access()` and `execFileAsync` wrapped in independent try/catch blocks. Confirmed no throw path from `initSubmodules`.
- **EACCES test**: Properly verifies non-fatal behavior — asserts active env and no submodule exec calls.
- **Symlink edge case**: `access()` follows symlinks. Broken symlink → ENOENT → skip (correct). Valid symlink → access succeeds → git reads resolved path (correct).
- **Config wire-up**: `initSubmodules` flows from `.archon/config.yaml` → `RepoConfig.worktree` → `WorktreeCreateConfig` → `createWorktree()` without any missing mapping.
- **Scope**: Minimal and focused — one config field, one method, one call site, six tests.

### Verdict

QA round 2: CLEAN — ready for review.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: packages/isolation/src/providers/worktree.ts [hunk 1]
```diff
@@ -919,6 +924,34 @@ export class WorktreeProvider implements IIsolationProvider {
...
+        ['-C', worktreePath, 'submodule', 'update', '--init', '--recursive'],
+        { timeout: 120000 }
+      );
+      getLog().info({ worktreePath }, 'worktree.submodules_initialized');
+    } catch (error) {
+      const err = error as Error;
+      // Non-fatal: the worktree is still usable without submodules.
+      // Log a warning so the user can diagnose if workflows fail.
+      getLog().warn({ err, worktreePath }, 'worktree.submodule_init_failed');
+    }
```
[the remark this answers is not in the document]
Comment (halindrome) [reply packages/isolation/src/providers/worktree.ts:1063]: Already addressed in commit 85ce222 (QA round 1 fix). The `access()` call is now inside its own `try/catch` — ENOENT returns silently, non-ENOENT (e.g., EACCES) logs a warning and returns without throwing. The `directoryExists` helper is no longer used. Both the access check and the exec are independently non-fatal.

**Comment by Wirasm** [comment]
Hey @halindrome — thanks for this, it closes a real gap and fits naturally with the recent worktree-primitive hardening series (#1198, #1206, #1212). I pushed two small follow-ups directly to your branch (`maintainerCanModify: true` — feel free to revert or amend if you disagree with the direction).

### 1. Default flipped from opt-in to auto-detect

Repos with `.gitmodules` present now have submodules initialized by default. Rationale: the `.gitmodules` probe is already a cheap `access()` check that short-circuits for non-submodule repos, so there's no cost to making it on-by-default. Defaulting to off means users with submodule monorepos (like yours) hit empty-dir state on first run — exactly the footgun this feature exists to prevent. Git itself takes the same stance with `submodule.recurse` / `clone --recurse-submodules`.

Explicit opt-out stays available via `worktree.initSubmodules: false` for users who don't want the fetch cost.

### 2. Submodule init failure now throws instead of warning

This one is the Archon convention more than a universal truth: [CLAUDE.md](../blob/main/CLAUDE.md#engineering-principles) calls out *\"Prefer throwing early with a clear error for unsupported or unsafe states — never silently swallow errors\"*. The prior behavior (`log.warn` + return) produces a worktree that *looks* healthy but has empty submodule dirs — which is the same silent-wrong-state class of bug the recent worktree PRs have been closing (#1186, #1198, #1206). A user who opted in to submodule init needs to know when it didn't actually happen.

The thrown error flows through `classifyIsolationError`, which I extended with a dedicated pattern that surfaces the opt-out path in the error message:

> **Error:** Submodule initialization failed. Check credentials and network access to submodule remotes, or set `worktree.initSubmodules: false` in `.archon/config.yaml` to opt out if submodules are not needed for your workflows.

EACCES on the `.gitmodules` read itself stays non-fatal (reading our own file is distinct from the git op; double-erroring adds no value).

### Validation

- `bun run type-check` — clean across all packages
- `bun --filter @archon/isolation test` — 228 tests pass (worktree provider + errors + others)
- `bun run lint` — zero warnings
- `bun run format:check` — clean

### Context on the broader class

Your four-issue root-cause mapping (#1183, #1187, #1188, #1192) has been incredibly useful — it's directly shaped the four PRs we've merged. The submodule init gap is orthogonal to the identity/coordination class those addressed; your issue framing made that clean separation possible. #1192 (codebase identity derivation) remains the underlying primitive rearchitect that would retire most of the guard code; that's the next logical step and will be tracked separately.

**Comment by Wirasm** [comment]
## PR Review Summary (multi-agent)

Ran code-reviewer, docs-impact, pr-test-analyzer, silent-failure-hunter, and type-design-analyzer against the diff.

### Critical Issues
None.

### Important Issues

| Agent | Issue | Location |
|-------|-------|----------|
| code-reviewer + silent-failure-hunter | Non-`ENOENT` errors on the `.gitmodules` `access()` check (e.g. `EACCES`) are logged at `warn` and silently skip submodule init. With `initSubmodules` defaulting to `true`, a `.gitmodules`-present-but-unreadable worktree is returned as if submodules initialized — exactly the silent-broken-state this PR exists to prevent. The test explicitly encodes this behavior as correct. Recommend: treat non-`ENOENT` access errors as fatal (re-throw via the same classified error) or at minimum surface to the user, not just logs. | `packages/isolation/src/providers/worktree.ts` ~lines 1037-1049; test at `worktree.test.ts` ~line 1058 |
| type-design-analyzer | `initSubmodules?: boolean` with `@default true` violates the standard optional-property reading (`undefined` conventionally means `false`). Here `undefined` → `true`. Recommend either making it non-optional with an explicit default at the parse/merge boundary, or very prominent docs. Also: `WorktreeCreateConfig.initSubmodules` and `RepoConfig.worktree.initSubmodules` are structurally duplicated — drift risk. | `packages/isolation/src/types.ts:242-253`, `packages/core/src/config/config-types.ts:143-160` |
| docs-impact | New `worktree.initSubmodules` option is absent from the user-facing config reference. Since it's now default-on, users on repos with unreachable submodule remotes will silently break until they discover the opt-out. | `packages/docs-web/src/content/docs/reference/configuration.md` ~line 119 (worktree block) |

### Suggestions

| Agent | Suggestion | Location |
|-------|------------|----------|
| pr-test-analyzer | Timeout path (120s `execFileAsync` timeout) is untested. A short test asserting the rejection message still matches `/Submodule initialization failed/` and that `classifyIsolationError` returns opt-out guidance would close the loop. Criticality 6/10. | `worktree.test.ts` |
| pr-test-analyzer | "Initializes when explicitly opted in" test only counts calls, doesn't assert `--recursive` in args. Low effort. | `worktree.test.ts` ~lines 1000-1012 |
| type-design-analyzer | No Zod validation on the parsed config — contract is only enforced by TS. Convention for engine schemas is Zod-derived. | `config-types.ts` |

### Strengths
- `execFileAsync` used correctly (not `exec`).
- Inner git-command `catch` logs at `error` with full context and re-throws a classifiable error — no swallowing on the hot path.
- `classifyIsolationError` registers the new "submodule initialization failed" pattern with actionable opt-out guidance, and `isKnownIsolationError` is updated in sync.
- Fail-fast on git-command failure is correct and tested.
- Default-on behavior tested both via absent config and `undefined` field.
- EACCES branch is at least explicitly tested (just encoded as "acceptable silent-skip", which is the disagreement above).

### Verdict
**NEEDS FIXES** — the EACCES silent-skip is the only material issue; the rest are polish. No critical blockers.

### Recommended Actions
1. Decide on EACCES behavior: either re-throw as classified error (preferred — matches fail-fast guideline) or surface to user via platform message. Update the test accordingly.
2. Add `initSubmodules` to `packages/docs-web/src/content/docs/reference/configuration.md` worktree block.
3. Consider resolving the `?: boolean` + `@default true` mismatch (make non-optional with default, or document prominently).
4. Optional: add timeout-path test; assert `--recursive` in opt-in test.
